data-prepper-plugins/s3-source/src/main/java/com/amazon/dataprepper/plugins/source/codec/CsvCodec.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.source.codec;
+
+import com.amazon.dataprepper.model.annotations.DataPrepperPlugin;
+import com.amazon.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import com.amazon.dataprepper.model.event.Event;
+import com.amazon.dataprepper.model.log.JacksonLog;
+import com.amazon.dataprepper.model.record.Record;
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.dataformat.csv.CsvMapper;
+import com.fasterxml.jackson.dataformat.csv.CsvReadException;
+import com.fasterxml.jackson.dataformat.csv.CsvSchema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+@DataPrepperPlugin(name = "csv", pluginType = Codec.class, pluginConfigurationType = CsvCodecConfig.class)
+public class CsvCodec implements Codec {
+    private static final Logger LOG = LoggerFactory.getLogger(CsvCodec.class);
+    private final CsvCodecConfig config;
+
+    @DataPrepperPluginConstructor
+    public CsvCodec(final CsvCodecConfig config) {
+        Objects.requireNonNull(config);
+        this.config = config;
+    }
+
+    @Override
+    public void parse(final InputStream inputStream, final Consumer<Record<Event>> eventConsumer) throws IOException {
+        try (final BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream))) {
+            parseBufferedReader(reader, eventConsumer);
+        }
+    }
+
+    private void parseBufferedReader(final BufferedReader reader, final Consumer<Record<Event>> eventConsumer) throws IOException {
+        final CsvMapper mapper = createCsvMapper();
+        final CsvSchema schema;
+        if (config.isDetectHeader()) {
+            schema = createAutodetectHeaderCsvSchema();
+        }
+        else {
+            final int numberColumnsFirstLine = getNumberOfColumnsByMarkingBeginningOfInputStreamAndResettingReaderAfter(reader);
+            schema = createCsvSchemaFromConfig(numberColumnsFirstLine);
+        }
+
+        MappingIterator<Map<String, String>> parsingIterator = mapper.readerFor(Map.class).with(schema).readValues(reader);
+        try {
+            while (parsingIterator.hasNextValue()) {
+                readCsvLine(parsingIterator, eventConsumer);
+            }
+        } catch (Exception jsonExceptionOnHasNextLine) {
+            LOG.error("An Exception occurred while determining if file has next line ", jsonExceptionOnHasNextLine);
+        }
+    }
+
+    private int getNumberOfColumnsByMarkingBeginningOfInputStreamAndResettingReaderAfter(BufferedReader reader) throws IOException {
```

Comment:
minor: I think you could just call this `getNumberOfColumns` since the caller of this function doesn't need to care how it works